packages/block-library/src/template-part/edit/use-async-list.js

```diff
@@ -0,0 +1,82 @@
+/**
+ * WordPress dependencies
+ */
+import { useEffect, useReducer } from '@wordpress/element';
+import { createQueue } from '@wordpress/priority-queue';
+
+/**
+ * Returns the first items from list that are present on state.
+ *
+ * @param {Array} list  New array.
+ * @param {Array} state Current state.
+ * @return {Array} First items present iin state.
+ */
+function getFirstItemsPresentInState( list, state ) {
+	const firstItems = [];
+
+	for ( let i = 0; i < list.length; i++ ) {
+		const item = list[ i ];
+		if ( ! state.includes( item ) ) {
+			break;
+		}
+
+		firstItems.push( item );
+	}
+
+	return firstItems;
+}
+
+/**
+ * Reducer keeping track of a list of appended items.
+ *
+ * @param {Array}  state  Current state
+ * @param {Object} action Action
+ *
+ * @return {Array} update state.
+ */
+function listReducer( state, action ) {
+	if ( action.type === 'reset' ) {
+		return action.list;
+	}
+
+	if ( action.type === 'append' ) {
+		return [ ...state, action.item ];
+	}
+
+	return state;
+}
+
+/**
+ * React hook returns an array which items get asynchronously appended from a source array.
+ * This behavior is useful if we want to render a list of items asynchronously for performance reasons.
+ *
+ * @param {Array} list Source array.
+ * @return {Array} Async array.
+ */
+function useAsyncList( list ) {
```

Comment:
I carried this file over from the inserter.  Maybe there is a better way to re-use this?

Reply:
it almost seems like it should be in `components` or even `compose`. Not sure where we have been putting hooks

you might also be able to avoid using it for now since it was a pretty recent addition. But certainly not a bad idea to have up front to improve scalability

Reply:
What is the rationale behind having this, vs. wrapping each preview in an async mode container?

Reply:
I think the main difference is that it has some state to take advantage of when the list changes.  This compares whats on state to the new list, keeps the first matching items on state, and starts appending new items one at a time from there.

I assume the async mode containers would still have the same ordered appending behavior, but would be starting that async list from the first item every time regardless of whether or not it makes sense for the first items to go through the process again.

Reply:
I think the difference is that the async mode containers don't make the initial render async, and this does. Right @youknowriad?

Can we move it to the `compose` package?

Reply:
> Can we move it to the compose package?

I would definitely feel better about this is we had it in a package instead.  Hopefully it will make sense to move it to something like `compose` ?

Reply:
Yes, I think it's safe to do that now.

Reply:
Cool, I moved `useAsyncList` over to the `compose` package and imported it into the inserter and this block from there.  Similarly removed priority-queue from the packages' deps since its no longer used in them.  (6c23e01)